File: app/src/main/java/com/kaciula/archiman/data/local/database/UserLocalDataSource.java
package com.kaciula.archiman.data.local.database;

import com.kaciula.archiman.data.UserDataSource;
import com.kaciula.archiman.domain.entity.User;
import io.reactivex.Completable;
import io.reactivex.Observable;
import io.realm.Realm;
import io.realm.RealmResults;
import java.util.ArrayList;
import java.util.List;

public class UserLocalDataSource implements UserDataSource {

  @Override
  public Observable<List<User>> getUsers() {
    return Observable.fromCallable(() -> {
      Realm realm = Realm.getDefaultInstance();
      RealmResults<UserDb> results = realm.where(UserDb.class).findAll();

      List<User> users = new ArrayList<>(results.size());
      for (UserDb userDb : results) {
        User user = new User(userDb.accountId, userDb.name);
        users.add(user);
      }
      realm.close();
      return users;
    });
  }

  @Override
  public Completable createOrUpdateUsers(final List<User> users) {
    return Completable.fromAction(() -> {
      Realm realm = Realm.getDefaultInstance();
      realm.executeTransaction(realm1 -> {
        List<UserDb> usersDb = new ArrayList<>(users.size());
        for (User user : users) {
          UserDb userDb = new UserDb();
          userDb.accountId = user.getAccountId();
          userDb.name = user.getName();
          usersDb.add(userDb);
        }
        realm1.copyToRealmOrUpdate(usersDb);
      });
      realm.close();
    });
  }
}

File: app/src/main/java/com/kaciula/archiman/data/local/database/UserLocalDataSource.kt
package com.kaciula.archiman.data.local.database

import com.kaciula.archiman.data.UserDataSource
import com.kaciula.archiman.domain.entity.User
import io.reactivex.Completable
import io.reactivex.Observable
import io.realm.Realm
import java.util.*

class UserLocalDataSource : UserDataSource {

    override fun getUsers(): Observable<List<User>> {
        return Observable.fromCallable<List<User>> {
            val realm = Realm.getDefaultInstance()
            val results = realm.where(UserDb::class.java).findAll()

            val users = ArrayList<User>(results.size)
            for (userDb in results) {
                val user = User(userDb.accountId, userDb.name!!)
                users.add(user)
            }
            realm.close()
            users
        }
    }

    override fun createOrUpdateUsers(users: List<User>): Completable {
        return Completable.fromAction {
            val realm = Realm.getDefaultInstance()
            realm.executeTransaction { realm1 ->
                val usersDb = ArrayList<UserDb>(users.size)
                for ((accountId, name) in users) {
                    val userDb = UserDb()
                    userDb.accountId = accountId
                    userDb.name = name
                    usersDb.add(userDb)
                }
                realm1.copyToRealmOrUpdate(usersDb)
            }
            realm.close()
        }
    }
}
